runtime/vm/jfr.cpp

```diff
@@ -1152,6 +1162,145 @@ jfrDump(J9VMThread *currentThread, BOOLEAN finalWrite)
 	flushAllThreadBuffers(currentThread, finalWrite);
 	writeOutGlobalBuffer(currentThread, finalWrite);
 }
+
+static UDATA
+jfrTypeIDHashFn(void *key, void *userData)
+{
+	const J9JFRTypeID *entry = (const J9JFRTypeID *)key;
+
+	const U_16 length = J9UTF8_LENGTH(entry->className);
+	const U_8 *data = J9UTF8_DATA(entry->className);
+
+	/* sdbm algorithm */
+	UDATA hash = 0;
+
+	for (U_16 i = 0; i < length; i++) {
+		hash = data[i] + (hash << 6) + (hash << 16) - hash;
+	}
+
+	return hash;
+}
+
+static UDATA
+jfrTypeIDHashEqualFn(void *tableNode, void *queryNode, void *userData)
+{
+	const J9JFRTypeID *tableEntry = (const J9JFRTypeID *)tableNode;
+	const J9JFRTypeID *queryEntry = (const J9JFRTypeID *)queryNode;
+
+	return J9UTF8_EQUALS(tableEntry->className, queryEntry->className);
+}
+
+UDATA
+initializeJFRIDs(J9JavaVM *vm)
+{
+	PORT_ACCESS_FROM_JAVAVM(vm);
+	UDATA result = 0;
+
+	vm->jfrState.typeIDs = hashTableNew(OMRPORT_FROM_J9PORT(privatePortLibrary),
+								J9_GET_CALLSITE(),
+								0,
+								sizeof(J9JFRTypeID),
+								sizeof(J9JFRTypeID *),
+								0,
+								J9MEM_CATEGORY_CLASSES,
+								jfrTypeIDHashFn,
+								jfrTypeIDHashEqualFn,
+								NULL,
+								NULL);
+
+	if (NULL == vm->jfrState.typeIDs) {
+		result = 1;
+		goto done;
+	}
+
+	if (omrthread_monitor_init_with_name(&vm->jfrState.typeIDMonitor, 0, "JFR Type ID monitor")) {
+		result = 1;
+		goto done;
+	}
+
+	/* Value needs to be the same as jdk.jfr.internal.JVM.RESERVED_CLASS_ID_LIMIT. */
+	vm->jfrState.typeIDcount = 500;
+
+done:
+	return result;
+}
+
+static UDATA
+freeJFRIDStrings(void *entry, void *userData)
+{
+	J9JavaVM *vm = (J9JavaVM *)userData;
+	J9JFRTypeID *jfrTypeID = (J9JFRTypeID *)entry;
+
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	if (jfrTypeID->free) {
+		j9mem_free_memory((void *)jfrTypeID->className);
+		jfrTypeID->className = NULL;
+	}
+
+	return FALSE;
+}
+
+void
+shutdownJFRIDs(J9JavaVM *vm)
+{
+	if (NULL != vm->jfrState.typeIDs) {
+		hashTableForEachDo(vm->jfrState.typeIDs, freeJFRIDStrings, vm);
+		hashTableFree(vm->jfrState.typeIDs);
+		vm->jfrState.typeIDs = NULL;
+	}
+
+	if (NULL != vm->jfrState.typeIDMonitor) {
+		omrthread_monitor_destroy(vm->jfrState.typeIDMonitor);
+		vm->jfrState.typeIDMonitor = NULL;
+	}
+}
+
+/**
+ * Must not be holding on to VM access
+ */
+I_64
+getTypeIDUTF8(J9VMThread *currentThread, const J9UTF8 *className, BOOLEAN free)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	J9HashTable *typeIDTable = vm->jfrState.typeIDs;
+	I_64 result = -1;
+	J9JFRTypeID entry = {0};
+	J9JFRTypeID *jfrTypeID = &entry;
+	jfrTypeID->className = className;
+
+	Trc_VM_getTypeIDUTF8_Entry(currentThread, J9UTF8_LENGTH(className), J9UTF8_DATA(className));
+
+	Assert_VM_mustHaveVMAccess(currentThread);
+
+	internalReleaseVMAccess(currentThread);
```

Comment:
This is likely unnecessary unless there are other interactions between VM access (including exclusive) and the typeIDMonitor. As long as the ordering is consistent you should be OK (for the most part, treat acquiring VM access like a monitor enter for ordering purposes).